Objects/genobject.c

```diff
@@ -1876,21 +1876,20 @@ async_gen_athrow_send(PyAsyncGenAThrow *o, PyObject *arg)
     return NULL;
 
 check_error:
-    if (PyErr_ExceptionMatches(PyExc_StopAsyncIteration)) {
+    if (PyErr_ExceptionMatches(PyExc_StopAsyncIteration) ||
+            PyErr_ExceptionMatches(PyExc_GeneratorExit))
+    {
         o->agt_state = AWAITABLE_STATE_CLOSED;
         if (o->agt_args == NULL) {
             /* when aclose() is called we don't want to propagate
-               StopAsyncIteration; just raise StopIteration, signalling
-               that 'aclose()' is done. */
+               StopAsyncIteration or GeneratorExit; just raise
```

Comment:
As far as I understand, this will make `.aclose()` raise `StopIteration` in the usual case (when generator doesn't handle the `GeneratorExit`). What about `.athrow()`, how will it behave? Shouldn't their behavior be similar to that of `.close()` and `.throw()` for sync generators, where `.close()` swallows the `GeneratorExit` and `.throw()` doesn't?

Reply:
> As far as I understand, this will make .aclose() raise StopIteration in the usual case (when generator doesn't handle the GeneratorExit).

Keep in mind, that raising `StopIteration` from any of `a***()` methods just means that they are returning. They are all iterators.

> What about .athrow(), how will it behave?

The idea here:

1. `athrow()` and `aclose()` share one implementation.

2. `aclose()` is basically an `athrow(GeneratorExit)`.

3. When `aclose()` is called and the generator propagates `GeneratorExit`, it means that everything is fine and the `aclose()` call was successful.  We raise `StopIteration` to make `aclose()` awaitable return.

4. When `athrow()` is called we don't want to trap *any* exceptions at all.  So we raise `StopIteration` only if the asynchronous generator swallowed the thrown error and yielded (asynchronously).

This is a bit complex, I know. I explained how asynchronous generators work in detail here: https://www.python.org/dev/peps/pep-0525/#implementation-details

Reply:
Thanks, now I get it